### ISSUE TITLE
SPM Plugin support for Package.swift

### ISSUE THREAD

**Issue by apps4everyone** [tuist/tuist#6096]
### What problem or need do you have?

When moving from external Package dependencies to Package.swift there is currently no SPM Plugin support.

### macOS version

14.3.1

### Tuist version

5.6.0

### Xcode version

15.3

**Comment by AnthonyMDev** [comment]
+1 to this. Would really like to see this implemented.

We want to use the [Benchmarks](https://github.com/ordo-one/package-benchmark/tree/main) library, but this functions as an SPM executable target and uses an SPM plugin.

**Comment by YIshihara11201** [comment]
Hi! I'd like to work on this issue.

**Comment by fortmarek** [comment]
> Hi! I'd like to work on this issue.

This would be amazing @YIshihara11201 💜 Let us know if you need any help/extra pointers 😌

**Comment by YIshihara11201** [comment]
Hi @fortmarek !

I  took some time to go through the codebase to better understand this issue.

> A target dependency between the target that uses the build plugin, and the target that represents the build plugin (command line tool)
> A -load-plugin-executable flag in the Swift compiler flag build setting to point to the executable in the derived data directory.

For the second part, I roughly identified where it is handled:
```
◼︎ Locating the executable
  - GraphTraverser.allSwiftPluginExecutables
◼︎ Setting the -load-plugin-executable flag for the swiftc command
  - ConfigGenerator.swiftMacrosDerivedSettings
◼︎ Placing the executable in a path where it can be loaded via -load-plugin-executable
  - BuildPhaseGenerator.generateBuildPhase
```

As for the first part, I feel I need to better understand the problem this issue refers to.
Specifically, what does the phrase "moving from external Package dependencies to Package.swift" actually mean in this context? 
**_It would be very helpful if you could clarify what specific steps or changes this phrase refers to._**

**Comment by pepicrft** [comment]
@YIshihara11201 there are two ways to integrate packages: - The official one where Xcode takes care of everything and the one where Tuist converts everything into Xcode targets.

This issue is about supporting the latter, so a built plugin is an executable. You would need to create a target in a project that is an executable, and then those built plugins need to be compiled before the target that needs them. So you need to declare a target dependency, and then you need to adjust the build setting in the target that depends on the built plugin to tell the Swift compiler to use the executable. 

Does that help?

**Comment by Yusuke-Ishihara-sp** [comment]
@pepicrft 
Thank you for the message.

I've reviewed your advice alongside the codebase, and I'm currently organizing what needs to be done before diving into the implementation.
Before I start coding in earnest, would you mind reviewing my current understanding and the planned approach? I'd really appreciate your feedback.

**Comment by YIshihara11201** [comment]
@pepicrft  
Thank you for the response.

I've reviewed your advice alongside the codebase, and I'm currently organizing what needs to be done before diving into the implementation.  

It might take me a couple of days, as I need to wrap my head around how the SPM plugin system itself works, and a few other things.

Before I start coding in earnest, would you mind reviewing my current understanding and planned approach? I'd really appreciate your feedback.

**Comment by YIshihara11201** [comment]
@pepicrft 
Just to quickly confirm one thing.
When managing a plugin via SPM, It's valid to write `Package.swift` like this:

```swift
// swift-tools-version:5.6
import PackageDescription

let package = Package(
    name: "MyApp",
    targets: [
        // ...,

        .executableTarget(
            name: "My-Benchmark",
            dependencies: [
                .product(name: "Benchmark", package: "package-benchmark"),
            ],
            path: "Benchmarks/My-Benchmark",
            plugins: [
                .plugin(name: "BenchmarkPlugin", package: "package-benchmark"),
            ]
        ),
    ],
    dependencies: [
        .package(url: "https://github.com/ordo-one/package-benchmark", .upToNextMajor(from: "1.4.0")),
    ]
)
```
Based on that, I’d like to check if my understanding is correct regarding the current state of Tuist.
- As far as I can tell, the following features are not yet implemented in Tuist:
  - Generating a plugin target (i.e., an executable) from the .plugin(...) declaration.
  - Linking that generated plugin target to any executableTarget that depends on it.
  - Adding the -load-plugin-executable compiler flag when compiling targets that depend on a plugin.

Is this understanding correct?

**Comment by YIshihara11201** [comment]
@pepicrft 
Let me check some word definitions in this [message](https://github.com/tuist/tuist/issues/6096#issuecomment-2995627336).
## Q1

> The official one where Xcode takes care of everything 

Is this referring to the process to `add .package(url:) to Project's property of packages in Project.swift` as following ?
```swift
let project = Project(
    name: "Framework",
    packages: [
        .package(url: "https://github.com/ordo-one/package-benchmark", .upToNextMajor(from: "1.4.0")),
    ],
  // ...
```
For this case, after `tuist install` `tuist generate`:  plugin is treated as a native package dependency. 

<image src="https://github.com/user-attachments/assets/68e39d6e-8149-40ca-9e93-bf7d12b86adb" width=350>




## Q2 

> the one where Tuist converts everything into Xcode targets.

Does this mean we must convert SPM plugin dependency `defined in Package.swift` into Xcode Projects, create targets, and  set up dependencies between those targets ?

**Comment by fortmarek** [comment]
> Is this referring to the process to add .package(url:) to Project's property of packages in Project.swift as following ?

Right, that use case is already supported.

> Does this mean we must convert SPM plugin dependency defined in Package.swift into Xcode Projects, create targets, and set up dependencies between those targets ?

That's exactly right. An important piece is the [PackageInfoMapper](https://github.com/tuist/tuist/blob/main/cli/Sources/TuistLoader/SwiftPackageManager/PackageInfoMapper.swift) that does the majority of the mapping from SwiftPM definition to the Tuist definition. This mapping currently does not support SwiftPM plugins. If you search for `macro` in that file, you'll see the important bits for converting a macro into the Tuist definition.

**Comment by YIshihara11201** [comment]
Hi!

I’m doing some experiments and would like to clarify one point.

I selected [SwiftLintPlugin](https://github.com/lukepistrol/SwiftLintPlugin/blob/main/Package.swift#L26) as a sample and noticed that the plugin depends on `SwiftLintBinary.artifactbundle.zip` instead of an xcframework.

Does this mean we would need to introduce a new architecture to handle this pattern properly? Or is there an existing approach that I might have overlooked?

**Comment by fortmarek** [comment]
> Does this mean we would need to introduce a new architecture to handle this pattern properly? Or is there an existing approach that I might have overlooked?

You will need to unzip the bundle when generating the project – and then grab the `.xcframework` that should be inside the archive. It might be good to start with that in a separate PR _or_ choose a different plugin that doesn't depend on a zipped binary target.

**Comment by YIshihara11201** [comment]
I would like to confirm the expected behavior.

With SPM’s native behavior, when a plugin is added as a dependency, it is automatically listed under Run Build Tool Plug-ins in Xcode.

<img src="https://github.com/user-attachments/assets/5512018d-a6a9-4c60-a852-7f641378ed02" width="850">

When supporting SPM plugins in Tuist, is it expected that the same behavior should occur?

**Comment by fortmarek** [comment]
The build tool plug-in support for the XcodeProj-based integration will look more like macros where we use macOS executables and build phases:

<img width="1652" height="1009" alt="Image" src="https://github.com/user-attachments/assets/9e64b2f9-5cd2-4cb6-a42b-24238cb903ec" />

This screenshot is taken from `multiplatform_app_with_swift_macros` fixture that might be good for you to inspect before doing any work on this one. Let me know if you need any help btw 🙂

**Comment by YIshihara11201** [comment]
@fortmarek 
> This screenshot is taken from multiplatform_app_with_swift_macros fixture that might be good for you to inspect before doing any work on this one. 

I’ve studied the macro workflow and am now applying that approach to SPM plugin integration. In this PR (#7990), I added a test fixture to verify the behaviour and left several comments. I’d appreciate your review to confirm whether I’m moving in the right direction. 🙏
